apps/progress/progress.c

```diff
@@ -0,0 +1,359 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include "progress.h"
+
+void insert(struct failure *f)
+{
+    struct node *node = malloc(sizeof(struct node));
+    node->val = f;
+    node->next = NULL;
+
+    if (head == NULL) {
+        head = node;
+        tail = node;
+    } else {
+        tail->next = node;
+        tail = node;
+    }
+}
+
+void free_list(struct node **head_ref)
+{
+    struct node *tmp;
+    struct node *head = *head_ref;
+    struct failure *f;
+    while (head != NULL) {
+        tmp = head;
+        head = head->next;
+        f = tmp->val;
+        free(f->name);
+        if (f->is_dynamic) free(f->error);
+        free(f);
+        free(tmp);
+    }
+}
+
+int main(int argc, char *argv[])
+{
+    for (int i = 1; i < argc; i++) {
+        char *arg = argv[i];
+        int len = strlen(arg);
+        for (int j = 0; j < len; j++) {
+            switch (arg[j]) {
+                case 'i':
+                    mode = INLINE;
+                    break;
+                case 'c':
+                    mode = CHARACTER;
+                    break;
+                case 'l':
+                    mode = LINES;
+                    break;
+                case 's':
+                    mode = SUMMARY;
+                    break;
+                case 't':
+                    indent++;
+                    break;
+            }
+        }
+    }
+
+    if (mode == SUMMARY) indent = 0;
+
+    report_start();
+    while (1) {
+        char v[1];
```

Comment:
Character-by-character is fine for now; I just did it that way in Virgil because I was too lazy to make multi-character reads from System.* reliably detect the number of available characters available and read only those (soooo rusty with Unix pipes). Can you add a TODO to do a multi-character read here?

Reply:
Just implemented multicharacter using `getline`.